backend/fms_core/template_importer/row_handlers/library_qc/library_qc.py

```diff
@@ -0,0 +1,116 @@
+
+
+
+from fms_core.models.instrument_type import InstrumentType
+from fms_core.utils import convert_concentration_from_nm_to_ngbyul
+from fms_core.services.sample import update_sample, get_sample_from_container
+from fms_core.services.process_measurement import create_process_measurement
+from fms_core.services.property_value import create_process_measurement_properties
+from fms_core.services.library import set_library_size
+from fms_core.template_importer.row_handlers._generic import GenericRowHandler
+
+INSTRUMENT_PROPERTIES = ['Library Quality Instrument', 'Library Quantity Instrument']
+QC_PLATFORM = "Quality Control"
+
+class LibraryQCRowHandler(GenericRowHandler):
+    def __init__(self):
+        super().__init__()
+
+    def process_row_inner(self, sample_container, measures, process, process_measurement, process_measurement_properties):
+         # Get the library sample that was checked
+        source_sample_obj, self.errors['container'], self.warnings['container'] = \
+            get_sample_from_container(barcode=sample_container['container_barcode'], coordinates=sample_container['container_coord'])
+
+        if source_sample_obj is None:
+            self.errors['sample_source'] = 'Library sample for QC was not found at the specified container.'
+
+        if not source_sample_obj.is_library:
+            self.errors['sample_source'] = f'The sample at the specified location is not a library: ${source_sample_obj.name}, id: ${source_sample_obj.id}'
+
+        # volume
+        if measures['initial_volume'] is None and measures['measured_volume'] is None:
+            self.errors['measured_volume'] = 'Either measured or initial volume must be specified.'
+
+        if measures['volume_used'] is None:
+            self.errors['volume_used'] = 'Volume used must be specified.'
+
+        if measures['volume_used'] < 0:
+            self.errors['Volume used should be a positive number.']
+
+        final_volume = measures['measured_volume'] if measures['measured_volume'] is not None else measures['initial_volume']
+        final_volume -= measures['volume_used']
+
+        # concentration
+        if measures['concentration_nm'] is not None and measures['concentration_uL'] is not None:
+            self.warning['concentration'] = 'The concentration should be specified in nM or uL, not both. nM will be ignored.'
```

Comment:
We try not to make choices for the users. The ultimate goal in Freezeman is not to make the user life easier it is to make sure the data is correctly saved. The user QoL comes second. In this case maybe the user actually converted the nM (real measurement) into ng/ul which would have come at a conversion cost. Maybe he trust his nM value more than the converted value ... In any case we want to leave the decision in his hand. It would be better to raise an error and refuse the template and let the user decide. If you want to get more fancy you can actually convert the nm to ng/ul and check if both values and equal if they are you can downgrade the error to a warning to let the user know they are doing more work than necessary.